Add solid mechanics tutorial (#353)
4C-multiphysics/4C: doc/readthedocs/tut_solid_preexo.rst

.. figure:: figures/tut_solid_flow_chart.jpg
   :alt: plasticity models flow chart
   :width: 800px
   :align: center

   Capabilities of plasticity models available in 4C

Preprocessing
----------------

Due to the symmetry of this model, we may restrict our analysis to one half in longitudinal direction.
In addition, one can see that the structure has a rotational symmetry.
Since axisymmetric elements are not available in |FOURC|, a 3D structure is modelled with double symmetry in both lateral directions, so that only one eighth of the structure is left.
The final FE model is shown here.

.. figure:: figures/tut_solid-mesh.jpg
   :alt: tutorial_solid mesh
   :width: 500px
   :align: center

   Mesh of one eighth of the tensile bar

Symmetry boundary conditions are applied to the three symmetry planes, denoted with ``XSYMM, YSYMM, ZSYMM`` in the above figure.
The top surface is exposed with a Dirichlet boundary condition with linearly increasing displacement in y-direction.

To this end, we model 200 time steps spanning from :math:`t_0 = 0` s to :math:`t_1 = 1` s with a time step of :math:`\Delta t = 0.005` s.

> Assuming nonlinear kinematics (large strain theory) and plasticity with isotropic hardening following an exponential curve,

Author: I added the equation, knowing that the model used is not the simplest one ;-)

Large strain/large deformation IMO depends on the mechanics school you are affiliated to.

Note that you need to build the ``pre_exodus`` command before.
You'll find the build information for ``pre_exodus`` in the :ref:`setup section <custom_target_specifiers>` or in the :ref:`workflow/preprocessing section <preprocessing>`.

The boundary condition file ``tutorial_solid.bc`` contains the element information and boundary conditions.
The element type is a quadratic (27 node) 3D solid element(Solid HEX27) with nonlinear kinematics, that is, including a large strain formulation.
The boundary conditions include three symmetry conditions (based on the surfaces defined in Cubit) and one loading condition.
Note that only surface conditions are given; the lines that belong to two surfaces automatically get the combined boundary conditions.
The loading condition has a function number (1) included for the y-direction, and also gets a tag: ``TAG monitor_reaction``.
The tag is for an extra output of the total forces, which are printed in a csv file, if the output is requested.

.. literalinclude:: tutorial_solid.bc
   :linenos:

All control parameters for the material, solvers, loading function, output information are contained in ``tutorial_solid.head``.

> Particularly, the output of the forces is requested by a particular section, ``--IO/MONITOR STRUCTURE DBC``, which controls accuracy and frequency of the output.

Reviewer: ```suggestion
Particularly, the output of the forces is requested by a particular section, ``IO/MONITOR STRUCTURE DBC``, which controls accuracy and frequency of the output.
```

Let's stop prefixing sections with `--`. Even for dat, this is kind of an implementation detail; for yaml/json, it is just wrong.

The end of the simulation is defined either when the maximum time or when the number of increments is reached, whatever comes earlier.
Note that the number of time steps is currently set to 10, in order to keep the simulation time short for the CI/CD tests.
If you want to run the whole simulation including necking of the specimen, you have to change the parameter to at least ``NUMSTEP 200``.

Besides output and solver details, the material and any functions are defined in the header.
For the material, we choose a large strain plasticity model, which is named ``MAT_Struct_PlasticNlnLogNeoHooke`` with parameters, which might represent a high strength Aluminum.
Last but not least, a function is defined in this file, to which the Dirichlet boundary condition refers.

.. literalinclude::  tutorial_solid.head
   :lines: 64-68
   :lineno-start: 64

Simulation
-----------

After the file ``tutorial_solid.dat`` has been created successfully using the ``pre_exodus`` command shown above, the simulation is run on a single processor by the command

::

    <build-dir>/4C tutorial_solid.dat tutorial_solid

Since the solver output is quite lengthy, one might want to redirect it into a file.

> For running the simulation on several processors, the ``mpirun -np <n_cpu>`` is used, on a cluster one probably has to use a batch script for running the simulation.

Reviewer: ```suggestion
For running the simulation with several processes, the ``mpirun -np <n_cpu>`` is used.
```

Post processing
----------------

The VTK results are collected in the directory ``tutorial_solid-vtk-files/``.
If they shall be opened by Paraview, the data file ``tutorial_solid-structure.pvd`` contains the complete collection of files, so only this file is to be opened, all other files are then read automatically.

> Using the filter ``Warp by vector`` and setting the last time step, one may get a picture like this one:

Reviewer: ```suggestion
Using the filter ``Warp by vector`` with the ``displacement`` vector, one may obtain the deformed geometry:
```
